### PR TITLE
Comments Redesign: Improve back navigation

### LINKED ISSUE #19972: Comments: Improve the Back button landing position
Currently the Back button simply brings back to the previous list.
We could improve this by scrolling to the last actioned comment.

Props to @rachelmcr for finding this!

### LINKED ISSUE #20300: Comments: Save scroll position after visiting another comment view then hitting back
In the comments section, when clicking into the post specific view or a single comment view, let's save our scroll position. Currently when we hit the back button, the window scrolls to top, and we lose our position. This can be disruptive for folks trying to go through a backlog of comments to moderate, since they need to scroll down and search for the last comment they read.

cc @apeatling

### PR DESCRIPTION
**Pull request by Copons**
Fix #19972
Partly fix #20300

## The issue

The Comments Management section has 3 different views: the site list, the post list, and the comment list.
While users can load all of these views directly, what happens when I navigate from site to post to comment and then back again?
Currently nothing really: the Back button of both post and comment views are hardcoded to redirect to the site view.
This is clearly not good.

An easy answer would be to simply use the browser's back feature. Though, this would likely mean losing the previous view's scroll position, which is quite important when working with long lists.
Let's keep track of which comment we clicked to open the post view!

## The fix

For conciseness, this PR only touches the site → post → site flow. If this is approved, I'll apply it to the comment view as well (it requires creating a new lengthy-ish component).

So: when we click on the `CommentPostLink`, instead of simply moving to the post view, we perform a bunch of new things:
- First and foremost: we scroll to the top of the page to avoid changing view while staying scrolled down.
- Then we replace the current browser's history state by adding a `#comment-${ commentId }` hash.
- Then we move to the post view.

The post view Back button is now wired up with `page` back function instead of having an arbitrary `href`.
Clicking on it, will in fact go to the previous page, which was modified with the `#comment-${ commentId }` hash.

Upon landing back to the site view, giving an `id="comment-${ commentId }` to the `Comment` elements is unfortunately not enough because of how React renders components.
(I still gave the `id` to the `Comment` elements for likely overkill backward compatibility.)

Instead, we wait for the `Comment` component to be fully loaded and, if we haven't scrolled yet, we obtain its node's `offsetTop` value, and (animate) scroll to that.
If for any reasons this scroll is not quite right, we force another quick scroll, which should be enough.

## Testing instructions

1

- Open `/comments`.
- Click on the post link of a comment "below the fold".
- Click on the `HeaderCake`'s Back button.
- The site view should be scrolled to the previously clicked comment.

2

- Do the same with the browser's Back button: it should have the same behaviour.

3

- Open directly a post view (e.g. `/comments/all/example.wordpress.com/12345`).
- The back button should now redirect to the site view (e.g. `/comments/all/example.wordpress.com`).

4

- Open `/comments`.
- Click on a post link of a comment "below the fold".
- Change list filter (e.g. from Approved to Spam).
- Click on either the `HeaderCake` or browser Back button.
- It should open the previous list filter (e.g. from Spam back to Approved).
- Click Back again.
- It should open the site view scrolled to the previously clicked comment.

### CODE REVIEW

File: client/my-sites/comments/comment/index.jsx [hunk 1]
```diff
@@ -38,13 +39,38 @@ export class Comment extends Component {
...
+
+		if (
+			! hasScrolled &&
+			! isLoading &&
+			!! window &&
+			`#comment-${ commentId }` === window.location.hash
+		) {
+			const commentNode = ReactDom.findDOMNode( this.commentCard );
+			const commentOffsetTop = commentNode.offsetTop - 56;
+			window.requestAnimationFrame( () =>
```
Comment (gwwar) [inline client/my-sites/comments/comment/index.jsx]: Should we give the page a chance to load all comments before scrolling down?

File: client/my-sites/comments/comment/index.jsx [hunk 2]
```diff
@@ -38,13 +39,38 @@ export class Comment extends Component {
...
+		const { hasScrolled } = this.state;
+
+		if (
+			! hasScrolled &&
+			! isLoading &&
+			!! window &&
+			`#comment-${ commentId }` === window.location.hash
+		) {
+			const commentNode = ReactDom.findDOMNode( this.commentCard );
+			const commentOffsetTop = commentNode.offsetTop - 56;
```
Comment (gwwar) [inline client/my-sites/comments/comment/index.jsx]: Maybe move the magic number here into a constant

File: client/my-sites/comments/comment/index.jsx (hunk 1 above)
Comment (gwwar) [reply client/my-sites/comments/comment/index.jsx]: Looks like we're queueing multiple `scrollTo` actions. I'd recommend changing the `hasScrolled` state into `isScrolling`. In order for this to work though we'll want to delay this call until comments have loaded.

eg:

```
if (
 			! isScrolling &&
 			! isLoading &&
 			!! window &&
 			`#comment-${ commentId }` === window.location.hash
		) {
     this.setState( { 'isScrolling', true } );
     //... queue scrollTo
}
```

Comment (gwwar) [reply client/my-sites/comments/comment/index.jsx]: Also note that scrollTo uses RAF internally so I think we can avoid that here.

**Comment by gwwar** [comment]
Thanks @Copons I like the approach here. I think we're seeing some oddness with the animation here since multiple handlers are being added. I added some notes on how to avoid this, but it requires us being able to tell when all comments are loaded.

Working on https://github.com/Automattic/wp-calypso/issues/20301 first may make this problem easier to tackle.

**Comment by Copons** [comment]
## Changes

`CommentList` gains a new `state.loadedComments` array populated by a method passed down to each `Comment`s, which trigger it when `props.isLoading` turns from `true` to `false`.

Once `CommentList.state.loadedComments` contains all the comments displayed in the page, we know that all these comments are in fact loaded.
We pass this information to all `Comment`s, so that the one `Comment` identified by the `#comment-${ commentId }` hash string will finally trigger the scroll.

In other words, I've moved both the "is loading" and "has scrolled" logic into the `CommentList`.

I've also taken care of a bunch of issues related to how Page.js and `window.history` handle paths, query params, and hash strings.

File: client/my-sites/comments/comment/index.jsx [hunk 3]
```diff
@@ -44,7 +53,40 @@ export class Comment extends Component {
...
+			const commentOffsetTop = commentNode.offsetTop - COMMENT_SCROLL_TOP_MARGIN;
+			scrollTo( {
+				x: 0,
+				y: commentOffsetTop,
+				duration: 1,
+				onComplete: () => {
+					if ( commentOffsetTop !== window.scrollY ) {
+						window.scrollTo( 0, commentOffsetTop );
+					}
+					onScrollToComment();
```
Comment (gwwar) [inline client/my-sites/comments/comment/index.jsx]: I believe this should still moved above the scrollTo

Comment (gwwar) [reply client/my-sites/comments/comment/index.jsx]: Casually testing, we're triggering the scroll 4 times on single page render! The `isScrolling` bit will prevent us from calling `ScrollTo` those additional times. eg `! hasScrolledToComment && ! isScrolling`

screen shot 2017-12-07 at 12 08 42 pm

**Comment by gwwar** [comment]
To follow up on Kirk's question @apeatling how do you feel about proposed behavior?

**Comment by apeatling** [comment]
I can't get 3. to work at all, if I load directly and go back I'm not seeing the site view. I'm using the live branch, can someone confirm it works there?

**Comment by kwight** [comment]
@apeatling works for me; I fired up the branch, then copied the post link to a new tab for a fresh post-specific load. Clicking the header cake back arrow took me to the All view.

**Comment by gwwar** [comment]
There are still bugs to work out, but I was curious of what you thought about the described behavior @apeatling

**Comment by apeatling** [comment]
Okay, the problem is I'm misreading, 3. says "The back button" so I was trying that. It's working fine for me. Number 4 doesn't make sense to me, it wasn't what I expected.

**Comment by Copons** [comment]
@apeatling my bad on number 4: I knew that the last changes I pushed last week broke it, but forgot to update the testing instructions accordingly.
Though, I've just added the back logic to the comment view as well, and it should now work as intended. I'll update the testing instructions as soon as everything works as expected.

File: client/my-sites/comments/comment/comment-link.jsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		const path = get( window, 'history.state.path' );
+
+		const newPath =
+			-1 !== path.indexOf( '#' )
+				? path.replace( /[#].*/, `#comment-${ commentId }` )
+				: `${ path }#comment-${ commentId }`;
+
+		window.history.replaceState( { ...window.history.state, path: newPath }, null );
+
+		page( `/comment/${ siteSlug }/${ commentId }` );
```
Comment (rodrigoi) [inline client/my-sites/comments/comment/comment-link.jsx]: I think that allowing components to have direct access to `page` is problematic. The navigation middleware has a `NAVIGATE` action that just calls `page( path )`. Maybe we can replace this with `dispatch( { type: NAVIGATE, path } )` if it doesn't break anything else.

File: client/my-sites/comments/comment/comment-post-link.jsx [hunk 5]
```diff
@@ -17,26 +19,60 @@ import { getSiteComment } from 'state/selectors';
...
+		const path = get( window, 'history.state.path' );
+
+		const newPath =
+			-1 !== path.indexOf( '#' )
+				? path.replace( /[#].*/, `#comment-${ commentId }` )
+				: `${ path }#comment-${ commentId }`;
+
+		window.history.replaceState( { ...window.history.state, path: newPath }, null );
+
+		page( `/comments/${ status }/${ siteSlug }/${ postId }` );
```
Comment (rodrigoi) [inline client/my-sites/comments/comment/comment-post-link.jsx]: Same as https://github.com/Automattic/wp-calypso/pull/20567/files#r156117524

File: client/my-sites/comments/comment-list/index.jsx [hunk 6]
```diff
@@ -153,6 +174,13 @@ export class CommentList extends Component {
 
 	isCommentSelected = commentId => !! find( this.state.selectedComments, { commentId } );
 
+	isLoadedAll = () => {
```
Comment (vindl) [inline client/my-sites/comments/comment-list/index.jsx]: Nitpick naming suggestion: `hasLoadedAllComments`.

File: client/my-sites/comments/comment/comment-link.jsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	};
+
+	render() {
+		const { commentId, commentUrl, isBulkMode, siteSlug } = this.props;
+
+		return isEnabled( 'comments/management/comment-view' ) ? (
```
Comment (vindl) [inline client/my-sites/comments/comment/comment-link.jsx]: We can clean this up now since we removed the feature flag.

File: client/my-sites/comments/comment/comment-link.jsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+		if ( ! window ) {
+			return;
+		}
+		event.preventDefault();
+		window.scrollTo( 0, 0 );
+
+		const { commentId, siteSlug } = this.props;
+		const path = get( window, 'history.state.path' );
+
+		const newPath =
```
Comment (vindl) [inline client/my-sites/comments/comment/comment-link.jsx]: Why do we have to perform path replacement here instead of just appending the `#comment-${ commentId }` to `href` attribute in render?

File: client/my-sites/comments/comment/comment-post-link.jsx [hunk 9]
```diff
@@ -17,26 +19,60 @@ import { getSiteComment } from 'state/selectors';
...
-} ) => (
-	<div className="comment__post-link">
-		{ ! isPostTitleLoaded && <QueryPosts siteId={ siteId } postId={ postId } /> }
+export class CommentPostLink extends PureComponent {
+	static propTypes = {
+		commentId: PropTypes.number,
+		isBulkMode: PropTypes.bool,
+	};
+
+	handleClick = event => {
```
Comment (vindl) [inline client/my-sites/comments/comment/comment-post-link.jsx]: It would be good if we could avoid duplicating the code in this function.

**Comment by gwwar** [comment]
@copons did you want to close this one out or rebase with the latest changes?

**Comment by Copons** [comment]
Replaced by #21617
